feat: Add GOMEMLIMIT to k8s templates

nberlee opened the pull request:
# Pull Request
fixes #7874
<!--
## Note to the Contributor

We encourage contributors to go through a proposal process to discuss major changes.
Before your PR is allowed to run through CI, the maintainers of Talos will first have to approve the PR.
-->

## What? (description)
This commit integrates the GOMEMLIMIT environment variable into the shipped k8s manifests when resources.limits.memory are defined. It is set to 95% of the memory limit to optimize the performance of the Go garbage collector, mitigating the risk of OOMKills in containerized environments.
## Why? (reasoning)
see issue #7874 :
Addressing inefficiencies of Go's garbage collector in constrained memory environments, this PR introduces the GOMEMLIMIT environment variable to k8s manifests with defined resources.limits.memory. This optimizes garbage collection frequency, reducing OOMKill risk. The GOMEMLIMIT is set to 95% of the memory limit, ensuring efficient garbage collection while accommodating undetected memory usages.

## Acceptance

Please use the following checklist:

- [X] you linked an issue (if applicable)
- [x] you included tests (if applicable)
- [X] you ran conformance (`make conformance`)
- [X] you formatted your code (`make fmt`)
- [X] you linted your code (`make lint`)
- [ ] you generated documentation (`make docs`)
- [X] you ran unit-tests (`make unit-tests`)

> See `make help` for a description of the available targets.

smira commented:
I wonder if we should do something like that `kube-apiserver` (?), but Talos doesn't define any limits iirc by default.

We were looking into doing same for Talos core components, currently looking into optimizing memory usage.

nberlee commented:
Yes, I just wrote something for api+controller+scheduler, will test it later today and add it to the PR

nberlee commented:
@smira I've added GOMEMLIMIT to api-server, controller-manager and scheduler if resources.limits are defined in the machine config.

I also stumbled on the controller-manager + scheduler not doing anything with resources set in the machine config. I fixed that in a different commit. If you want me to merge the commit, create an issue and or make a different PR for that, please say so.

smira commented on internal/app/machined/pkg/controllers/k8s/control_plane.go:
```diff
@@ -167,6 +167,7 @@ func NewControlPlaneControllerManagerController() *ControlPlaneControllerManager
 					ExtraArgs:            cfgProvider.Cluster().ControllerManager().ExtraArgs(),
 					ExtraVolumes:         convertVolumes(cfgProvider.Cluster().ControllerManager().ExtraVolumes()),
 					EnvironmentVariables: cfgProvider.Cluster().ControllerManager().Env(),
+					Resources:            convertResources(cfgProvider.Cluster().APIServer().Resources()),
```
should be `ControllerManager().Resources()`? (and similar below?)

nberlee replied:
yes :flushed: 

Also didn't noticed during testing as I just also copy pasted the limits :(

fixed now

smira commented:
> I also stumbled on the controller-manager + scheduler not doing anything with resources set in the machine config. I fixed that in a different commit. If you want me to merge the commit, create an issue and or make a different PR for that, please say so.

that's a great find, this PR is good as well!

smira commented:
/ok-to-test

smira reviewed (approved):
added one more unit-test, thank you!
